slides/k8s/netpol.md

```diff
@@ -0,0 +1,267 @@
+# Network policies
+
+- Namespaces help us to *organize* resources
+
+- Namespaces do not provide isolation
+
+- By default, every pod can contact every other pod
+
+- By default, every service accepts traffic from anyone
+
+- If we want this to be different, we need *network policies*
+
+---
+
+## What's a network policy?
+
+A network policy is defined by the following things.
+
+- A *pod selector* indicating which pods it applies to
+
+  e.g.: "all pods in namespace `blue` with the label `zone=internal`"
+
+- A list of *ingress rules* indicating which inbound traffic is allowed
+
+  e.g.: "TCP connections to ports 8000 and 8080 coming from pods with label `zone=dmz`,
+  and from the external subnet 4.42.6.0/24, except 4.42.6.5"
+
+- A list of *egress rules* indicating which outbound traffic is allowed
+
+A network policy can provide ingress rules, egress rules, or both.
+
+---
+
+## How do network policies apply?
+
+- A pod can be "selected" by any number of network policies
+
+- If a pod isn't selected by any network policy, then its traffic is unrestricted
+
+  (In other words: in the absence of network policies, all traffic is allowed)
+
+- If a pod is selected by at least one network policy, then all traffic is blocked ...
+
+  ... unless it is explicitly allowed by one of these network policies
+
+---
+
+class: extra-details
+
+## Traffic filtering is flow-oriented
+
+- Network policies deal with *connections*, not individual packets
+
+- Example: to allow HTTP (80/tcp) connections to pod A, you only need an ingress rule
+
+  (You do not need a matching egress rule to allow response traffic to go through)
+
+- This also stands for UDP traffic
+
+  (Allowing DNS traffic can be done with a single rule)
+
+- Network policy implementations use stateful connection tracking
+
+---
+
+## Pod-to-pod traffic
+
+- Connections from pod A to pod B have to be allowed by both pods:
+
+  - pod A has to be unrestricted, or allow the connection as an *egress* rule
+
+  - pod B has to be unrestricted, or allow the connection as an *ingress* rule
+
+- As a consequence: if a network policy restricts traffic going from/to a pod,
+  <br/>
+  the restriction cannot be overridden by a network policy selecting another pod
+
+- This prevents an entity with access to namespace A (but no access to namespace B)
+  from adding network policies giving them access to namespace B
```

Comment:
This raises a question in my mind: does having "access" to, say, write files _in_ a namespace also allow an entity to _adjust policies_ for that namespace? I'm suddenly confused about "user" vs "admin".

Reply:
"Having access to a namespace" means (inb4 pedantry!) "being authenticated as a user or system account that has authorization to manage resources (in that context, including at least network policies) in that namespace".

(That was quite a mouthful, right :D)

If I try to read deeper in your question, I think you're asking "what about user access vs. admin access? Arguably, users shouldn't be able to manage network policies!". But that decision is completely up to the organization; i.e. they might decide "each team/individual/PR gets their own namespace, and can do whatever they want in it, including managing network policies". Or they might decide, "we only have a limited number of namespaces, each with its custom set of network policies that are closely managed by a dedicated team." (These are only two stances among many possible, of course.)

I didn't want to dive too deep into these details at this juncture, because there is a whole chapter on RBAC in the new content.

Now that I've written all this prose, I'm thinking: "Perhaps instead of _access_, I should write in full _permission to manage network policies_?"

> This prevents an entity managing network policies in namespace A (but without permission to do so in namespace B) from adding network policies giving them access to namespace B

How does that sound?